ddev/src/ddev/cli/meta/scripts/upgrade_python.py

```diff
@@ -20,57 +21,200 @@ def upgrade_python(app: Application, version: str):
+        app.display_warning("Documentation files have not been updated. Please modify them manually.")
+
+    tracker.display()
+
+    if tracker.errors:  # no cov
+        app.abort()
+
+
+def update_ci_files(app: Application, new_version: str, old_version: str, tracker: ValidationTracker):
+    for file in (app.repo.path / ".github" / "workflows").glob("*.yml"):
         changed = False
+        content = file.read_text()
 
-        for env in test_config.get('envs', {}).values():
-            default_python = env.get('python', '')
-            if default_python == old_version:
-                env['python'] = version
-                tracker.success()
+        for pattern in ("python-version: '{}'", 'PYTHON_VERSION: "{}"', "'{}'"):
+            if pattern.format(old_version) in content:
+                content = content.replace(pattern.format(old_version), pattern.format(new_version))
                 changed = True
 
-            for variables in env.get('matrix', []):
-                pythons = variables.get('python', [])
-                for i, python in enumerate(pythons):
-                    if python == old_version:
-                        pythons[i] = version
-                        tracker.success()
-                        changed = True
-
-            for overrides in env.get('overrides', {}).get('matrix', {}).get('python', {}).values():
-                for override in overrides:
-                    pythons = override.get('if', [])
-                    for i, python in enumerate(pythons):
-                        if python == old_version:
-                            pythons[i] = version
-                            tracker.success()
-                            changed = True
-
         if changed:
-            config_file.write_text(tomlkit.dumps(test_config))
+            file.write_text(content)
+            tracker.success()
+
+
+def update_ddev_template_files(app: Application, new_version: str, old_version: str, tracker: ValidationTracker):
+    for check_type in ("check", "jmx", "logs"):
+        folder_path = (
+            app.repo.path
+            / 'datadog_checks_dev'
+            / 'datadog_checks'
+            / 'dev'
+            / 'tooling'
+            / 'templates'
+            / 'integration'
+            / check_type
+            / '{check_name}'
+        )
+        pyproject_file = folder_path / 'pyproject.toml'
+        changed = False
 
-    if app.repo.name == 'core':
-        constant_file = app.repo.path / 'ddev' / 'src' / 'ddev' / 'repo' / 'constants.py'
+        if pyproject_file.is_file():
+            content = pyproject_file.read_text()
+
+            for pattern in ('requires-python = ">={}"', "Programming Language :: Python :: {}"):
+                if pattern.format(old_version) in content:
+                    content = content.replace(pattern.format(old_version), pattern.format(new_version))
+                    changed = True
+
+            if changed:
```

Comment:
A slightly more direct approach would be to just compare the content:

```python
old_content = new_content = pyproject_file.read_text()
# for loop over the patterns where we change `new_content` if needed
if old_content != new_content:
    # write the changes
```

If you don't care as much about efficiency you could even dump the contents of the file every time even if they didn't change. For a script like this that would be fine imo, we only run it a few times per year.

Same applies to updating CI files. It also applies to `pyproject.toml` if you go the "inefficient" route :)

Reply:
I thought about that, but we keep track of the number of modifications using the tracker. I agree that's not completely needed, but I like to have it so I'll keep it that way if that works for you

Reply:
I see, the tracker part requires us to check for changes. We can still drop the extra boolean flag though, right?

Reply:
I aded the boolean because I wanted to keep track of the number of files that are modified, not the number of modified lines. Do you think it's better to have the total number of modifications instead?

Reply:
Hmm, the snippet in my first comment also only checks the total file contents, not individual lines.

Reply:
🤦 sorry, I was thinking about my code, not your snippet. Updated!

Reply:
Done! 👍